<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=utf-8">
<meta id=Generator content="Microsoft Word 15 (filtered)">
<title>CONSULATE GENERAL OF CANADA</title>
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:Helvetica;
	panose-1:0 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:Courier;
	panose-1:0 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:SimSun;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:DengXian;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
@font-face
	{font-family:Cambria;
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:"Lucida Grande";
	panose-1:2 11 6 0 4 5 2 2 2 4;}
@font-face
	{font-family:Tahoma;
	panose-1:2 11 6 4 3 5 4 4 2 4;}
@font-face
	{font-family:"\@SimSun";
	panose-1:2 1 6 0 3 1 1 1 1 1;}
@font-face
	{font-family:"\@DengXian";
	panose-1:2 1 6 0 3 1 1 1 1 1;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
h1
	{mso-style-link:"Heading 1 Char";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:18.0pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-18.0pt;
	line-height:150%;
	background:white;
	font-size:14.0pt;
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	font-weight:bold;}
h2
	{mso-style-link:"Heading 2 Char";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:64.4pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-36.0pt;
	line-height:150%;
	background:white;
	font-size:12.0pt;
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	font-weight:bold;}
h3
	{mso-style-link:"Heading 3 Char";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:70.9pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-35.45pt;
	line-height:150%;
	background:white;
	font-size:11.0pt;
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	font-weight:bold;}
h4
	{mso-style-link:"Heading 4 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	border:none;
	padding:0cm;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-weight:normal;}
h5
	{mso-style-link:"Heading 5 Char";
	margin-top:16.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-weight:normal;}
h6
	{mso-style-link:"Heading 6 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	color:#943634;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-weight:normal;}
p.MsoHeading7, li.MsoHeading7, div.MsoHeading7
	{mso-style-link:"Heading 7 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	color:#943634;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-style:italic;}
p.MsoHeading8, li.MsoHeading8, div.MsoHeading8
	{mso-style-link:"Heading 8 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	text-transform:uppercase;
	letter-spacing:.5pt;}
p.MsoHeading9, li.MsoHeading9, div.MsoHeading9
	{mso-style-link:"Heading 9 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-style:italic;}
p.MsoToc1, li.MsoToc1, div.MsoToc1
	{margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;
	text-transform:uppercase;
	font-weight:bold;
	text-decoration:underline;}
p.MsoToc2, li.MsoToc2, div.MsoToc2
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;
	font-variant:small-caps;
	font-weight:bold;}
p.MsoToc3, li.MsoToc3, div.MsoToc3
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;
	font-variant:small-caps;}
p.MsoToc4, li.MsoToc4, div.MsoToc4
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoToc5, li.MsoToc5, div.MsoToc5
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoToc6, li.MsoToc6, div.MsoToc6
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoToc7, li.MsoToc7, div.MsoToc7
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoToc8, li.MsoToc8, div.MsoToc8
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoToc9, li.MsoToc9, div.MsoToc9
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoCommentText, li.MsoCommentText, div.MsoCommentText
	{mso-style-link:"Comment Text Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	font-size:10.0pt;
	font-family:"Cambria",serif;}
p.MsoHeader, li.MsoHeader, div.MsoHeader
	{mso-style-link:"Header Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoFooter, li.MsoFooter, div.MsoFooter
	{mso-style-link:"Footer Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoCaption, li.MsoCaption, div.MsoCaption
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:9.0pt;
	font-family:"Cambria",serif;
	text-transform:uppercase;
	letter-spacing:.5pt;}
p.MsoTitle, li.MsoTitle, div.MsoTitle
	{mso-style-link:"Title Char";
	margin-top:25.0pt;
	margin-right:0cm;
	margin-bottom:15.0pt;
	margin-left:0cm;
	text-align:center;
	border:none;
	padding:0cm;
	font-size:22.0pt;
	font-family:"Cambria",serif;
	color:#632423;
	text-transform:uppercase;
	letter-spacing:2.5pt;}
p.MsoSubtitle, li.MsoSubtitle, div.MsoSubtitle
	{mso-style-link:"Subtitle Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:28.0pt;
	margin-left:0cm;
	text-align:center;
	font-size:9.0pt;
	font-family:"Cambria",serif;
	text-transform:uppercase;
	letter-spacing:1.0pt;}
p.MsoDate, li.MsoDate, div.MsoDate
	{mso-style-link:"Date Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
a:link, span.MsoHyperlink
	{color:blue;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:#954F72;
	text-decoration:underline;}
strong
	{color:#943634;
	letter-spacing:.25pt;}
em
	{text-transform:uppercase;
	letter-spacing:.25pt;
	font-style:normal;}
p
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman",serif;}
pre
	{mso-style-link:"HTML Preformatted Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:10.0pt;
	font-family:Courier;}
p.MsoCommentSubject, li.MsoCommentSubject, div.MsoCommentSubject
	{mso-style-link:"Comment Subject Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	font-weight:bold;}
p.MsoAcetate, li.MsoAcetate, div.MsoAcetate
	{mso-style-link:"Balloon Text Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:8.0pt;
	font-family:"Tahoma",sans-serif;}
span.MsoPlaceholderText
	{color:gray;}
p.MsoRMPane, li.MsoRMPane, div.MsoRMPane
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoListParagraph, li.MsoListParagraph, div.MsoListParagraph
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoListParagraphCxSpFirst, li.MsoListParagraphCxSpFirst, div.MsoListParagraphCxSpFirst
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoListParagraphCxSpMiddle, li.MsoListParagraphCxSpMiddle, div.MsoListParagraphCxSpMiddle
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoListParagraphCxSpLast, li.MsoListParagraphCxSpLast, div.MsoListParagraphCxSpLast
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.MsoTocHeading, li.MsoTocHeading, div.MsoTocHeading
	{margin-top:24.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:18.0pt;
	margin-bottom:.0001pt;
	text-indent:-18.0pt;
	line-height:115%;
	page-break-after:avoid;
	background:white;
	font-size:14.0pt;
	font-family:"Calibri Light",sans-serif;
	color:#2F5496;
	text-transform:uppercase;}
p.MediumGrid21, li.MediumGrid21, div.MediumGrid21
	{mso-style-name:"Medium Grid 21";
	mso-style-link:"Medium Grid 2 Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
span.Heading1Char
	{mso-style-name:"Heading 1 Char";
	mso-style-link:"Heading 1";
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	background:white;
	font-weight:bold;}
span.Heading2Char
	{mso-style-name:"Heading 2 Char";
	mso-style-link:"Heading 2";
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	background:white;
	font-weight:bold;}
span.Heading3Char
	{mso-style-name:"Heading 3 Char";
	mso-style-link:"Heading 3";
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	background:white;
	font-weight:bold;}
span.Heading4Char
	{mso-style-name:"Heading 4 Char";
	mso-style-link:"Heading 4";
	font-family:"Times New Roman",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.5pt;}
span.Heading5Char
	{mso-style-name:"Heading 5 Char";
	mso-style-link:"Heading 5";
	font-family:"Times New Roman",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.5pt;}
span.Heading6Char
	{mso-style-name:"Heading 6 Char";
	mso-style-link:"Heading 6";
	font-family:"Times New Roman",serif;
	color:#943634;
	text-transform:uppercase;
	letter-spacing:.5pt;}
span.Heading7Char
	{mso-style-name:"Heading 7 Char";
	mso-style-link:"Heading 7";
	font-family:"Times New Roman",serif;
	color:#943634;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-style:italic;}
span.Heading8Char
	{mso-style-name:"Heading 8 Char";
	mso-style-link:"Heading 8";
	font-family:"Times New Roman",serif;
	text-transform:uppercase;
	letter-spacing:.5pt;}
span.Heading9Char
	{mso-style-name:"Heading 9 Char";
	mso-style-link:"Heading 9";
	font-family:"Times New Roman",serif;
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-style:italic;}
span.TitleChar
	{mso-style-name:"Title Char";
	mso-style-link:Title;
	font-family:"Times New Roman",serif;
	color:#632423;
	text-transform:uppercase;
	letter-spacing:2.5pt;}
span.SubtitleChar
	{mso-style-name:"Subtitle Char";
	mso-style-link:Subtitle;
	font-family:"Times New Roman",serif;
	text-transform:uppercase;
	letter-spacing:1.0pt;}
span.MediumGrid2Char
	{mso-style-name:"Medium Grid 2 Char";
	mso-style-link:"Medium Grid 21";}
p.ColorfulList-Accent11, li.ColorfulList-Accent11, div.ColorfulList-Accent11
	{mso-style-name:"Colorful List - Accent 11";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.ColorfulList-Accent11CxSpFirst, li.ColorfulList-Accent11CxSpFirst, div.ColorfulList-Accent11CxSpFirst
	{mso-style-name:"Colorful List - Accent 11CxSpFirst";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.ColorfulList-Accent11CxSpMiddle, li.ColorfulList-Accent11CxSpMiddle, div.ColorfulList-Accent11CxSpMiddle
	{mso-style-name:"Colorful List - Accent 11CxSpMiddle";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.ColorfulList-Accent11CxSpLast, li.ColorfulList-Accent11CxSpLast, div.ColorfulList-Accent11CxSpLast
	{mso-style-name:"Colorful List - Accent 11CxSpLast";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:36.0pt;
	line-height:105%;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.ColorfulGrid-Accent11, li.ColorfulGrid-Accent11, div.ColorfulGrid-Accent11
	{mso-style-name:"Colorful Grid - Accent 11";
	mso-style-link:"Colorful Grid - Accent 1 Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:0cm;
	line-height:105%;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	font-style:italic;}
span.ColorfulGrid-Accent1Char
	{mso-style-name:"Colorful Grid - Accent 1 Char";
	mso-style-link:"Colorful Grid - Accent 11";
	font-family:"Times New Roman",serif;
	font-style:italic;}
p.LightShading-Accent21, li.LightShading-Accent21, div.LightShading-Accent21
	{mso-style-name:"Light Shading - Accent 21";
	mso-style-link:"Light Shading - Accent 2 Char";
	margin-top:8.0pt;
	margin-right:72.0pt;
	margin-bottom:10.0pt;
	margin-left:72.0pt;
	line-height:125%;
	border:none;
	padding:0cm;
	font-size:10.0pt;
	font-family:"Cambria",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.25pt;}
span.LightShading-Accent2Char
	{mso-style-name:"Light Shading - Accent 2 Char";
	mso-style-link:"Light Shading - Accent 21";
	font-family:"Times New Roman",serif;
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.25pt;}
span.SubtleEmphasis1
	{mso-style-name:"Subtle Emphasis1";
	font-style:italic;}
span.IntenseEmphasis1
	{mso-style-name:"Intense Emphasis1";
	text-transform:uppercase;
	letter-spacing:.5pt;
	font-style:italic;}
span.SubtleReference1
	{mso-style-name:"Subtle Reference1";
	font-family:"Calibri",sans-serif;
	color:#622423;
	font-style:italic;}
span.IntenseReference1
	{mso-style-name:"Intense Reference1";
	font-family:"Calibri",sans-serif;
	color:#622423;
	font-weight:bold;
	font-style:italic;}
span.BookTitle1
	{mso-style-name:"Book Title1";
	color:#622423;
	text-transform:uppercase;
	letter-spacing:.25pt;}
p.TOCHeading1, li.TOCHeading1, div.TOCHeading1
	{mso-style-name:"TOC Heading1";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:18.0pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-18.0pt;
	line-height:150%;
	background:white;
	font-size:14.0pt;
	font-family:"Lucida Grande",sans-serif;
	color:#222222;
	font-weight:bold;}
span.BalloonTextChar
	{mso-style-name:"Balloon Text Char";
	mso-style-link:"Balloon Text";
	font-family:"Tahoma",sans-serif;}
span.HeaderChar
	{mso-style-name:"Header Char";
	mso-style-link:Header;}
span.FooterChar
	{mso-style-name:"Footer Char";
	mso-style-link:Footer;}
p.StyleLinespacing15lines, li.StyleLinespacing15lines, div.StyleLinespacing15lines
	{mso-style-name:"Style Line spacing\:  1\.5 lines";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	line-height:120%;
	font-size:12.0pt;
	font-family:"Times New Roman",serif;}
p.ColorfulShading-Accent11, li.ColorfulShading-Accent11, div.ColorfulShading-Accent11
	{mso-style-name:"Colorful Shading - Accent 11";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
p.ColorfulShading-Accent12, li.ColorfulShading-Accent12, div.ColorfulShading-Accent12
	{mso-style-name:"Colorful Shading - Accent 12";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Cambria",serif;}
span.HTMLPreformattedChar
	{mso-style-name:"HTML Preformatted Char";
	mso-style-link:"HTML Preformatted";
	font-family:Courier;}
p.Abstract-Text, li.Abstract-Text, div.Abstract-Text
	{mso-style-name:Abstract-Text;
	mso-style-link:"Abstract-Text Char";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-justify:inter-ideograph;
	line-height:11.0pt;
	font-size:9.0pt;
	font-family:Helvetica;}
span.Abstract-TextChar
	{mso-style-name:"Abstract-Text Char";
	mso-style-link:Abstract-Text;
	font-family:Helvetica;}
span.UnresolvedMention1
	{mso-style-name:"Unresolved Mention1";
	color:gray;
	background:#E6E6E6;}
span.doc
	{mso-style-name:doc;}
p.EquationDisplay, li.EquationDisplay, div.EquationDisplay
	{mso-style-name:"Equation Display";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
span.DateChar
	{mso-style-name:"Date Char";
	mso-style-link:Date;}
p.para-first, li.para-first, div.para-first
	{mso-style-name:para-first;
	mso-style-link:"para-first Char";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-justify:inter-ideograph;
	line-height:11.0pt;
	font-size:8.0pt;
	font-family:"Times New Roman",serif;}
span.para-firstChar
	{mso-style-name:"para-first Char";
	mso-style-link:para-first;
	font-family:"Times New Roman",serif;}
p.Text3, li.Text3, div.Text3
	{mso-style-name:"Text 3";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:36.0pt;
	text-align:justify;
	text-justify:inter-ideograph;
	line-height:150%;
	background:white;
	font-size:11.0pt;
	font-family:"Lucida Grande",sans-serif;
	color:#222222;}
span.CommentTextChar
	{mso-style-name:"Comment Text Char";
	mso-style-link:"Comment Text";}
span.CommentSubjectChar
	{mso-style-name:"Comment Subject Char";
	mso-style-link:"Comment Subject";
	font-weight:bold;}
p.RefHead, li.RefHead, div.RefHead
	{mso-style-name:"Ref Head";
	margin-top:18.0pt;
	margin-right:0cm;
	margin-bottom:2.5pt;
	margin-left:0cm;
	line-height:12.0pt;
	font-size:10.0pt;
	font-family:Helvetica;
	font-weight:bold;}
p.RefText, li.RefText, div.RefText
	{mso-style-name:"Ref Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:11.35pt;
	margin-bottom:.0001pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-11.35pt;
	line-height:9.0pt;
	font-size:7.0pt;
	font-family:"Times New Roman",serif;}
.MsoChpDefault
	{font-size:10.0pt;
	font-family:"Cambria",serif;}
 /* Page Definitions */
 @page WordSection1
	{size:612.0pt 792.0pt;
	margin:49.7pt 22.6pt 58.3pt 72.0pt;}
div.WordSection1
	{page:WordSection1;}
 /* List Definitions */
 ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>

</head>

<body lang=EN-CA link=blue vlink="#954F72">

<div class=WordSection1>

<p class=MsoTocHeading align=center style='margin-left:0cm;text-align:center;
text-indent:0cm'><b><span style='font-size:12.0pt;line-height:115%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Human Protein Variant Effect Map Imputation
Toolkit User Guide</span></b></p>

<p class=MsoToc1><span
class=MsoHyperlink><span lang=EN-US><a href=#_Toc520809761>1.<span
lang=EN-US style='font-size:12.0pt;line-height:105%;text-transform:none;
font-weight:normal;text-decoration:none'>    </span>Upload and impute<span
style='display:none'>  </span><span
style='display:none'>3</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535294">1.1<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Select
Landscape File<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>3</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535295">1.1.1<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Raw Data
(Checkbox)<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>3</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc520809764">1.1.2<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Example –
UBE2I and BRCA1<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>3</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535297">1.2<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Select
Protein Fasta File<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>4</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535298">1.3<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Input
Uniprot ID<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>4</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535299">1.4<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Set
Parameters<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>4</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535300">1.4.1<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Filter low
quality missense variants<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>4</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535301">1.4.2<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Data
rescaling<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>5</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535302">1.4.3<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Standard
deviation regularization<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>5</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535303">1.4.4<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Auto
training variants quality cutoff<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535304">1.5<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Name Your
Session<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535305">1.6<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Email
Address<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc1><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc520809774">2.<span lang=EN-US style='font-size:12.0pt;line-height:
105%;text-transform:none;font-weight:normal;text-decoration:none'>    </span>View
Landscapes<span style='display:none'>  </span><span
style='display:none'>6</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535307">2.1<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Previously
imputed maps<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535308">2.2<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>View maps
by session ID<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535309">2.3<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>View maps
by Uniprot ID<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>6</span></a></span></span></p>

<p class=MsoToc1><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc520809777">3.<span lang=EN-US style='font-size:12.0pt;line-height:
105%;text-transform:none;font-weight:normal;text-decoration:none'>    </span>Downloads<span
style='display:none'>  </span><span
style='display:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535311">3.1<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
Weile et al. 2017<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535312">3.2<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
Data Template<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535313">3.3<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
Help Document<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535314">3.4<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
Supported Uniprot IDs<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc1><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535315">4.<span lang=EN-US style='font-size:12.0pt;line-height:"src/com/alphame/imputation/client/resource/imputation_help.html"
105%;text-transform:none;font-weight:normal;text-decoration:none'>    </span>Map
Viewer<span style='display:none'>  </span><span
style='display:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535316">4.1<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
CSV<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535317">4.2<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Download
Figures<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535318">4.3<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>PubMed
Link<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>7</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535319">4.4<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>View
options / Legends<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535320">4.4.1<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Original
Fitness<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535321">4.4.2<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Refined
Fitness<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535322">4.4.3<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Polyphen
Score<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535323">4.4.4<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>SIFT Score<span
style='display:none;text-decoration:none'>  </span><span
style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535324">4.4.5<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Provean
Score<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>8</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535325">4.4.6<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Allele
Frequency<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>9</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535326">4.5<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Map
Visualization<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>9</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535327">4.5.1<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Accessible
Surface Area (ASA) track<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>9</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535328">4.5.2<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Secondary
structure (SS) track<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>9</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535329">4.5.3<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Pfam
domain track<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>9</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535330">4.5.4<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Variant
tracks<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>10</span></a></span></span></p>

<p class=MsoToc3><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535331">4.5.5<span lang=EN-US style='font-size:12.0pt;line-height:
105%;font-variant:normal !important;text-decoration:none'>    </span>Pop Up
Panel<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>10</span></a></span></span></p>

<p class=MsoToc1><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535332">5.<span lang=EN-US style='font-size:12.0pt;line-height:
105%;text-transform:none;font-weight:normal;text-decoration:none'>    </span>Appendix<span
style='display:none'>  </span><span
style='display:none'>10</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535333">5.1<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Raw data
format<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>10</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535334">5.2<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Processed
data format<span style='display:none;text-decoration:none'> </span><span style='display:none;text-decoration:none'>11</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535335">5.3<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>Imputation
result data format<span style='display:none;text-decoration:none'>  </span><span style='display:none;text-decoration:none'>11</span></a></span></span></p>

<p class=MsoToc2><span class=MsoHyperlink><span lang=EN-US><a
href="#_Toc527535336">5.4<span lang=EN-US style='font-size:12.0pt;font-variant:
normal !important;font-weight:normal;text-decoration:none'>    </span>References<span
style='display:none;text-decoration:none'> </span><span
style='display:none;text-decoration:none'>13</span></a></span></span></p>

<p class=MsoNormal><span lang=EN-US>&nbsp;</span></p>

<p class=MsoNormal style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
margin-left:0cm;line-height:150%'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif'>Welcome
to the&nbsp;Human Protein Variants Effect Map Imputation Toolkit User Guide. The
purpose of this toolkit is to impute, refine and visualize data from
multiplexed variant effect assays (MAVEs) to produce missense variant effect
(VE) maps. Specifically, we provide: 1) a front end web application (<span
class=MsoHyperlink><a href="http://impute.varianteffect.org/">http://impute.varianteffect.org/</a></span><span
class=MsoHyperlink><span style='text-decoration:none'>, source: </span><a
href="https://github.com/joewuca/imputation">https://github.com/joewuca/imputation</a></span>)
that allows users to upload their own MAVE data and visualize or download a
complete map of variant effects; and 2) a back end data processing service that
performs imputation and refinement. This user guide will focus on the front end
web application and proper and effective use of the toolkit. For users
interested in the back end data processing service, please see previous methods
in Weile et al, (Molecular Systems Biology 2017) and other methodological
references to be added here as the toolkit evolves to implement other methods.</span></p>

<p class=MsoNormal style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
margin-left:0cm;line-height:150%'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif'>The
user guide is organized into four sections based on the user interface (see the
purple boxes on the figure below): <b>(1) Upload and Impute:</b> for uploading input
MAVE data, metadata and parameters, imputing missing values, and refining
measured values. <b>(2) View Landscapes:</b> for selecting and visualizing the
finished variant effect maps (either from the user or selected from
previously-finished examples).</span><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'> </span><b><span lang=EN-US style='font-family:"Lucida Grande",sans-serif'>(3)
Downloads:</span></b><span lang=EN-US style='font-family:"Lucida Grande",sans-serif'>
for downloading related information such as data template and help files <b>(4)
Map viewer:</b> for visualizing and downloading finished variant effect maps.</span></p>

<p class=MsoNormal style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
margin-left:0cm;line-height:150%'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:red'>NOTE:  human protein variant effect maps (imputed or otherwise) are
research tools and should be appropriately validated before clinical use.</span></p>

<p class=MsoNormal style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
margin-left:0cm;line-height:150%'><span style='font-family:"Lucida Grande",sans-serif'>&nbsp;</span></p>

<p class=MsoNormal style='margin-top:6.0pt;margin-right:-3.6pt;margin-bottom:
6.0pt;margin-left:0cm;line-height:150%'><span lang=EN-US style='font-family:
"Lucida Grande",sans-serif;color:#222222'><img border=0 width=448 height=223
id="Picture 4" src="http://impute.varianteffect.org/images/imputation_help.fld/image001.jpg"></span></p>

<h1><a id="_Toc520809761">1.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; 
</span><span dir=LTR></span>Upload and impute</h1>

<h2><a id="_Toc527535294">1.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Select Landscape File</a></h2>

<h3><a id="_Toc527535295">1.1.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Raw Data (Checkbox)</a> </h3>

<p class=Text3>There are two supported data formats for the input protein MAVE
data:</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>Raw data format</b></p>

<p class=Text3 style='margin-left:72.0pt'>A raw landscape data format for human
protein variant effect map calculation and imputation, developed for raw MAVE
data from the TileSeq method (Welie <i>et.,al</i> 2017). Please refer to Section
5.1 for the details.</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>Processed data format</b></p>

<p class=Text3 style='margin-left:72.0pt'>The processed data format is a more
general data format can be used for MAVE data generated from a variety of experimental
approaches.  Please refer to Section 5.2 for the details.</p>

<h3><a id="_Toc520809764">1.1.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Example – UBE2I and BRCA1</a></h3>

<p class=Text3><span lang=EN-US><img border=0 width=54 height=19 id="Picture 1"
src="http://impute.varianteffect.org/images/imputation_help.fld/image002.jpg"></span></p>

<p class=Text3>Clicking the example button will initiate download of a zip file
with files needed for imputing the variant effect map of two example proteins (UBE2I;
the human SUMO conjugase enzyme that recognizes sumoylation sites on client
proteins and attaches the SUMO protein to these clients, data collected from
Welie <i>et.,al</i> 2017. BRCA1; the human breast cancer type 1 susceptibility
protein, data collected from Findlay <i>et.,al</i> 2018) will be downloaded to
your local disk, and the default parameters for UBE2I imputation will be loaded
as well. There are six files in the zip file:</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=PL
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=PL>rawData_UBE2I.txt (UBE2I raw
data)</span></p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=IT
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=IT>processedData_UBE2I.txt (UBE2I
processed data)</span></p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=IT
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=IT>P63279_UBE2I.fasta (UBE2I full
protein sequence)</span></p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=IT
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=IT>processedData_BRCA1-RING.txt (BRCA1
RING domain processed data)</span></p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=IT
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=IT>processedData_BRCA1-BRCT.txt (BRCA1
BRCT domain processed data)</span></p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span lang=IT
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><span lang=IT>P38398_BRCA1.fasta (BRCA1 full
protein sequence)</span></p>

<p class=Text3>Before you hit the “impute” button to run the example protein
imputation, please make sure you have loaded the raw/processed data and the fasta
file, and also have correctly checked the raw data checkbox if you are using
the raw data file as input (or left it unchecked if you are using the processed
data input file). The running time for the example imputation is expected to be
less than one minute. Please note clicking the example button will set the
session name and Uniprot ID for UBE2I imputation, to impute BRCA1 you need to
change the Uniprot ID to P38398 and also change your session name after loading
the BRCA1 related input files.</p>

<h2><a id="_Toc527535297">1.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Select Protein Fasta File</a></h2>

<p class=Text3 style='margin-left:28.4pt'>Please select the protein fasta file
that is consistent with your input data. For example, if your MAVE experiment
was done on the non-canonical protein variant, your input fasta file should
contain the protein sequence from this non-canonical variant.  </p>

<h2><a id="_Toc527535298">1.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Input Uniprot ID</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Please input the correct Uniprot ID corresponding
to your target protein to retrieve the precomputed protein features such as structure
features, Polyphen2 and SIFT scores. We have precomputed features for the most
of reviewed Uniprot human IDs in ClinVAR database (Currently ~3200 proteins). The
full list of supported Uniprot IDs can be download, please check Section 3.4
for details. If you submit a valid human Uniprot ID that is not in our list, you
will be given an option to request retrieval of protein features and notified
by email when this is complete.</span></p>

<h2><a id="_Toc527535299">1.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Set Parameters</a></h2>

<h3><a id="_Toc527535300">1.4.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Filter low quality missense variants</a></h3>

<p class=Text3>This option filters out input data for those missense variant
records with a quality score below the defined quality score cutoff: </p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>variant quality score cutoff</b></p>

<p class=Text3 style='margin-left:2.0cm'>The cutoff used for removing input data
for variants for which MAVE data does not surpass a specified quality score.</p>

<h3><a id="_Toc527535301">1.4.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Data rescaling</a></h3>

<p class=Text3>The data rescaling option rescales the fitness of each variant such
that the median of stop codon variants has fitness 0 after rescaling and the
median of synonymous variants has fitness 1 after rescaling. We set the variant
fitness score to 0 If the score is smaller than 0 after rescaling. If your
processed data is already rescaled to your preference, you should uncheck the
“Data rescaling” checkbox. If the box is checked, you can set up the following
two parameters:</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>Synonymous/nonsense variant quality
cutoff</b></p>

<p class=Text3 style='margin-left:2.0cm'>This cutoff is used for filtering out the
low quality synonymous and nonsense variants that you do not want to include
for median calculation.  The default value is 0.</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>Nonsense variant exclusion regions</b></p>

<p class=Text3 style='margin-left:2.0cm'>Although it is often assumed that
nonsense variants are severe null mutations, this is not necessarily the case,
especially for nonsense variants near the 3’ end of the original open reading
frame, i.e., near the wild-type stop codon.  You can define the regions of the
protein for which nonsense variants should be excluded in the median
calculation used for rescaling. The input format is “a:b” to exclude a single
region from position a to position b (e.g. to exclude nonsense variants in the
last 10% of the wild type open reading frame for a 100 amino acid protein, use
“90:100”. In rare cases it may be desired to exclude multiple regions.  This
can be done by concatenating inputs (e.g.  “a:b,c:d”).  The default value is “0:0”.</p>

<h3><a id="_Toc527535302">1.4.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Standard deviation regularization</a></h3>

<p class=Text3>This option regularizes the measured standard deviation (SD)
with a prior estimate of SD calculated from an overall regression of SD values
against fitness values when the number of replicates is small.  This Bayesian
regularization procedure is based on Baldi, P. &amp; Long <i>et al.,</i> 2001.</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol;color:black'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b><span style='color:black'>Number of replicates
required to skip standard deviation regularization</span></b></p>

<p class=Text3 style='margin-left:2.0cm'>When an insufficient number of
replicates is available to accurately measure SD, it is appropriate to use
regularization to smooth observed data and ‘nudge’ SD estimates towards prior expectation
for the SD. This parameter defines a minimum number of replicates, <i>k</i>. If
the user provides <i>j</i> observed replicates and this is lower than <i>k</i>,
a regularized estimate of <i>σ</i> is calculated as in Weile <i>et al.,</i>
2017, effectively giving weight <i>j</i> / <i>k</i> to the observed data.
Larger values of <i>k</i> give more weight on the prior SD estimate during
regularization. The default value for <i>k</i> is 8.</p>

<h3><a id="_Toc527535303">1.4.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Auto training variants quality cutoff</a></h3>

<p class=Text3>The auto training quality cutoff is determined by searching for
the training data cutoff that gives the best performance in terms of predicting
the test dataset (which consists of the top 20% of variants, ranked by quality
score). If this option is unchecked, then the cutoff must be set manually:</p>

<p class=Text3 style='margin-left:72.0pt;text-indent:-18.0pt'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span dir=LTR></span><b>Training variants quality cutoff</b></p>

<p class=Text3 style='margin-left:2.0cm'>Manually set the quality score cutoff
to remove low-quality variants from the imputation training variant examples.</p>

<h2><a id="_Toc527535304">1.5<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Name Your Session</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>You can name your current imputation
session with any string such as the protein name, datetime etc. The final
session ID is set as “[Uniprot ID][Name Your Session string]”. Once the
imputation is done, you can visualize the result on the map viewer and also
revisit your results through View Landscapes using your session ID.   </span></p>

<h2><a id="_Toc527535305">1.6<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Email Address</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>You can input your email address to
receive an email alert when the imputation job is done. The email will provide your
session ID and links to download results.</span></p>

<h1><a id="_Toc520809774">2.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span><span dir=LTR></span>View Landscapes</a></h1>

<h2><a id="_Toc527535307">2.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Previously imputed maps</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Select a map from previously imputed variant
effect maps provided using the list box.</span></p>

<h2><a id="_Toc527535308">2.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>View maps by session ID</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>To view your previously-imputed variant
effect maps, input your session ID and click View.</span></p>

<h2><a id="_Toc527535309">2.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>View maps by Uniprot ID</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>To view maps for other predictors
(Polyphen2, SIFT and Provean), input the Uniport ID of interest and click View.
Note this function is only available for the Uniprot IDs supported by our system
(See section 3.4). </span></p>

<h1><a id="_Toc520809777">3.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span><span dir=LTR></span>Downloads</a></h1>

<h2><a id="_Toc527535311">3.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download Weile et al. 2017</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
0cm;text-align:justify;text-justify:inter-ideograph;text-indent:28.4pt;
line-height:150%;background:white'><span style='font-size:11.0pt;line-height:
150%;font-family:"Lucida Grande",sans-serif;color:#222222'>The button links to
the script, the original and imputed maps from Weile et al., 2017.</span></p>

<h2><a id="_Toc527535312">3.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download Data Template</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>The input data templates (raw and
processed).  Please do not change the column names as these are used for
parsing the input file. </span></p>

<h2><a id="_Toc527535313">3.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download Help Document</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
0cm;text-align:justify;text-justify:inter-ideograph;text-indent:28.4pt;
line-height:150%;background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>This button downloads this document in pdf format.</span></p>

<h2><a id="_Toc527535314">3.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download Supported Uniprot IDs</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
31.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>This button downloads a full list of supported Uniprot IDs so
far. The Boolean value for each column indicates the availability of the
corresponding feature. </span></p>

<h1><a id="_Toc527535315">4.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span><span dir=LTR></span>Map Viewer</a></h1>

<h2><a id="_Toc527535316">4.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download CSV</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>Download the imputed result in CSV format (see Section 5.3 for a
description of each column).</span></p>

<h2><a id="_Toc527535317">4.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Download Figures</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;line-height:150%;background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>Two figures can be downloaded: one figure shows the original input
variant effect map with missing values and the other shows the finished
(imputed and refined) variant effect map. Each is provided in PDF format. When
the Download Figures button is clicked, two separate web pages should open, one
for each figure.  Some browsers may have settings such that the second webpage
is blocked.  In this case, please alter browser settings to enable popup windows
for this web application. The figures are similar but not identical to the
html-encoded images provided in the visualization window. </span></p>

<h2><a id="_Toc527535318">4.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>PubMed Link</a></h2>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
28.4pt;line-height:150%;background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>Link to the paper in which the previously imputed variant effect
map was published.</span></p>

<h2><a id="_Toc527535319">4.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>View options / Legends</a></h2>

<h3><a id="_Toc527535320">4.4.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Original Fitness</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
70.9pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Original Fitness option visualizes your
input variants (after rescaling. The original fitness score range is 0 to 2
with 1 indicating wild-type (WT) fitness.</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
42.55pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=183 height=44 id="Picture 3"
src="http://impute.varianteffect.org/images/imputation_help.fld/image003.jpg"></span></p>

<h3><a id="_Toc527535321">4.4.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Refined Fitness</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Refined Fitness option visualizes the
variants effect map after imputation and refinement. The refined fitness core range
is 0 to 2 with 1 indicating WT fitness.</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
43.65pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=183 height=44 id="Picture 6"
src="http://impute.varianteffect.org/images/imputation_help.fld/image003.jpg"></span></p>

<h3><a id="_Toc527535322">4.4.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Polyphen Score</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Polyphen Score option visualizes the
variants effect map using polyphen2 score (http://genetics.bwh.harvard.edu/pph2/).
The polyphen2 score range is 0 to 1 representing the estimated probability that
the variant has a damaging impact on protein function. </span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
43.65pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=184 height=44 id="Picture 13"
src="http://impute.varianteffect.org/images/imputation_help.fld/image004.jpg"></span></p>

<h3><a id="_Toc527535323">4.4.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>SIFT Score</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>SIFT Score option visualizes the
variants effect map using SIFT score (http://sift.jcvi.org/). The SIFT score range
is 0 to 1 representing the probability of being tolerated. Variants with
normalized probabilities less than 0.05 are predicted to be deleterious, those
greater than or equal to 0.05 are predicted to be tolerated.&nbsp;</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
43.65pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=184 height=44 id="Picture 12"
src="http://impute.varianteffect.org/images/imputation_help.fld/image005.jpg"></span></p>

<h3><a id="_Toc527535324">4.4.5<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Provean Score</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>Provean Score option visualises the
variants effect map using Provean score (</span><span class=MsoHyperlink><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif'><a
href="http://provean.jcvi.org/">http://provean.jcvi.org/</a></span></span><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
color:#222222'>). The Provean score range is -13 to 4. If the Provean score is
equal to or below a predefined threshold (-2.5), the protein variant is
predicted to have a &quot;deleterious&quot; effect. If the Provean score is
above the threshold, the variant is predicted to have a &quot;neutral&quot; effect.</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
43.65pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=184 height=44 id="Picture 9"
src="http://impute.varianteffect.org/images/imputation_help.fld/image006.jpg"></span></p>

<h3><a id="_Toc527535325">4.4.6<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Allele Frequency</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>The Allele Frequency option
visualizes the variants effect map using the -log10(MAF) score, where MAF is
minor allele frequency as scored in the GNOMAD database (http://gnomad.broadinstitute.org/).
The -log10 MAF score ranging from 0.3 to 10.</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
36.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=184 height=44 id="Picture 11"
src="http://impute.varianteffect.org/images/imputation_help.fld/image007.jpg"></span></p>

<h2><a id="_Toc527535326">4.5<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Map Visualization</a> </h2>

<h3><a id="_Toc527535327">4.5.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Accessible Surface Area (ASA) track</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>The ASA track shows the relative
Accessible Surface Area of each position of the protein, which is re</span><span
style='font-family:"Lucida Grande",sans-serif;color:#222222'>trieved from a PDB</span><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
color:#222222'> structure of the input protein, where available. The ASA score
ranges from 0 to 1, where 1 indicates the highest accessible surface area value.</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
36.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span lang=EN-US style='font-family:"Lucida Grande",sans-serif;
color:#222222'><img border=0 width=184 height=44 id="Picture 14"
src="http://impute.varianteffect.org/images/imputation_help.fld/image008.jpg"></span></p>

<h3><a id="_Toc527535328">4.5.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Secondary structure (SS) track</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>The </span><span style='font-family:
"Lucida Grande",sans-serif;color:#222222'>SS</span><span style='font-size:11.0pt;
line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'> track
shows the secondary structure for each position of the protein. The SS data is
retrieved using PsiPred (</span><span class=MsoHyperlink><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif'><a
href="http://bioinf.cs.ucl.ac.uk/psipred/">http://bioinf.cs.ucl.ac.uk/psipred/</a></span></span><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
color:#222222'>). More specifically, the letters C, H, and E indicate coiled
coil, alpha helix and beta sheet respectively.</span></p>

<h3><a id="_Toc527535329">4.5.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Pfam domain track</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>The Pfam track shows the Pfam domain
of the protein. The Pfam domain boundaries are retrieved from Pfam database (</span><span
class=MsoHyperlink><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif'><a href="https://pfam.xfam.org/">https://pfam.xfam.org/</a></span></span><span
style='font-size:11.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
color:#222222'>). Clicking the domain name in the track leads to the Pfam
webpage of that domain. </span></p>

<h3><a id="_Toc527535330">4.5.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Variant tracks</a></h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>There are 21 variant tracks,
representing the 20 possible amino acid values plus the stop codon.  The color
code and value for each entry of variant tracks depend on the current selected
view options. For wild type entries, where the amino acid track corresponds to
the wild type amino acid, the color code is YELLOW (this is not shown on the
fitness score legend but is by definition fitness 1). The small vertical line
in each entry represents the standard error of the fitness score. </span></p>

<h3><a id="_Toc527535331">4.5.5<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Pop Up Panel</a> </h3>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
72.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-size:11.0pt;line-height:150%;font-family:
"Lucida Grande",sans-serif;color:#222222'>When you mouse over each entry, you
will see a popup panel showing the detailed information of that entry for all
view options. </span></p>

<h1><a id="_Toc527535332">5.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span><span dir=LTR></span>Appendix</a> </h1>

<h2><a id="_Toc527535333">5.1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Raw data format</a></h2>

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='margin-left:33.75pt;border-collapse:collapse;border:none'>
 <tr style='height:20.6pt'>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt;height:20.6pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Column
  Name</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border:solid windowtext 1.0pt;
  border-left:none;padding:0cm 5.4pt 0cm 5.4pt;height:20.6pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Description</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>wt_aa</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The wild type residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>Pos</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The position of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>mut_aa</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The substituted residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>wt_codon</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The wild type codon of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>mut_codon</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The substituted codon of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>Annotation</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>NONSYN/SYN/STOP</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>nonselect</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>The variant allele frequency in
  variant pool under non-selective condition</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>select</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>The variant allele frequency in
  variant pool under selective condition</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>controlNS</span></p>
  </td>
  <td width=460 style='width:344.65pt;border-top:none;border-left:none;
  border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>The apparent variant allele frequency
  in control(wt) </span><span lang=EN-US style='font-size:10.0pt;line-height:
  105%;font-family:"Lucida Grande",sans-serif'>template </span><span
  lang=EN-US style='font-size:10.0pt;line-height:105%;font-family:"Lucida Grande",sans-serif'>under
  non-selective condition (a control in which all variants are sequencing
  errors)</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>controlS</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>The apparent variant allele frequency
  in control(wt) </span><span lang=EN-US style='font-size:10.0pt;line-height:
  105%;font-family:"Lucida Grande",sans-serif'>template </span><span
  lang=EN-US style='font-size:10.0pt;line-height:105%;font-family:"Lucida Grande",sans-serif'>under
  selective condition (a control in which all variants are sequencing errors)</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 style='width:114.85pt;border:solid windowtext 1.0pt;border-top:
  none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;line-height:150%'><span style='font-size:10.0pt;line-height:150%;
  font-family:"Lucida Grande",sans-serif;color:#222222'>replicate_id</span></p>
  </td>
  <td width=460 style='width:344.65pt;border-top:none;border-left:none;
  border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal><span lang=EN-US style='font-size:10.0pt;line-height:105%;
  font-family:"Lucida Grande",sans-serif'>The id for replicates. Please make
  sure different replicate has different replicate id.</span></p>
  </td>
 </tr>
</table>

<p class=MsoNormal>&nbsp;</p>

<h2><a id="_Toc527535334">5.2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Processed data format</a></h2>

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='margin-left:33.75pt;border-collapse:collapse;border:none'>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Column
  Name</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border:solid windowtext 1.0pt;
  border-left:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Description</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_ref</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The wild type residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_pos</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The position of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_alt</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The substituted residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>quality_score</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The quality score of the variant. For example, the allele
  frequency of the variant in non-selective condition.</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>num_replicates</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The number of replicates used to calculate the fitness score.</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_input</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The original fitness score of a variant (mean taken over all
  replicates)</span></p>
  </td>
 </tr>
 <tr>
  <td width=153 valign=top style='width:114.85pt;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_input_sd</span></p>
  </td>
  <td width=460 valign=top style='width:344.65pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The standard deviation of original fitness scores across all
  replicates.</span></p>
  </td>
 </tr>
</table>

<p class=MsoNormal>&nbsp;</p>

<h2><a id="_Toc527535335">5.3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>Imputation result data format</a></h2>

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='margin-left:33.75pt;border-collapse:collapse;border:none'>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Column
  Name</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border:solid windowtext 1.0pt;
  border-left:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p align=center style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;
  margin-left:0cm;text-align:center;line-height:150%'><span style='font-size:
  10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;color:#222222'>Description</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_ref</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The wild type residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_pos</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The position of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_alt</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The substituted residue of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>annotation</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>STOP/SYN/NONSYN</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>quality_score</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The quality score of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>num_replicates</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>Number of replicates </span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>pseudo_counts</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The pseudocount of the replicates used for standard deviation
  Bayesian regularization</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_input</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The original fitness score of a variant (mean taken over all
  replicates)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_input_sd</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The standard deviation of original fitness scores across all
  replicates.</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_org</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The rescaled fitness of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  lang=EN-US style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_sd_org</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The standard deviation of the rescaled variant fitness values</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The reversed and floored fitness_org </span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_sd</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The rescaled fitness standard deviation of the variant (same
  as fitness_org_sd)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_sd_prior</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The prior (background) fitness standard deviation of the
  variant, derived from regression (fitness_sd against fitness)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_sd_reg</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The regularized standard deviation of variant fitness</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_se_reg</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The regularized standard error of variant fitness</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_imputed</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The imputed fitness of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_imputed_se</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The standard error of the imputed variant fitness value (|fitness_imputed
  – fitness|)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_imputed_se_prior</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The prior estimate of standard error used in imputing the standard
  error of variant fitness, derived from regression (fitness_imputed_se against
  fitness_imputed) </span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_refine</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The refined fitness of the variant, achieved by combining
  fitness and fitness_imputed (a weighted average with weighting by inverse of
  estimated variance)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>fitness_refine_se</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The standard error estimate for the refined variant fitness
  value</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>Polyphen_score</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The polyphen2 score of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>Sift_score</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The SIFT score of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>Provean_score</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The Provean score of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>gnomad_af</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The Gnomad-derived minor allele frequency of the variant</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>asa_mean</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The accessible surface area of the variant position</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>aa_psipred</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The secondary structure of the SS segment starting from
  aa_position to ss_end_pos (C: Coiled coil, H: Alpha Helix, E: Beta Sheet).
  The value only exists on the STOP variants record to avoid redundancy as
  there are 22 records for each position (1 STOP, 1 SYN, 20 NONSYN)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>ss_end_pos</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The end position of the SS segment starting from aa_position</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>hmm_id</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The HMM id of the Pfam domain starting from aa_position to pfam_end_pos.
  The value only exists on the STOP variants record to avoid redundancy as
  there are 22 records for each position (1 STOP, 1 SYN, 20 NONSYN)</span></p>
  </td>
 </tr>
 <tr>
  <td width=151 valign=top style='width:4.0cm;border:solid windowtext 1.0pt;
  border-top:none;padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>pfam_end_pos</span></p>
  </td>
  <td width=488 valign=top style='width:365.7pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
  0cm;text-align:justify;text-justify:inter-ideograph;line-height:150%'><span
  style='font-size:10.0pt;line-height:150%;font-family:"Lucida Grande",sans-serif;
  color:#222222'>The end position of Pfam domain starting from aa_position</span></p>
  </td>
 </tr>
</table>

<p class=MsoNormal>&nbsp;</p>

<h2><a id="_Toc527535336">5.4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span><span dir=LTR></span>References</a></h2>

<p class=RefText style='margin-left:39.7pt;line-height:normal'><span
style='font-size:11.0pt;font-family:"Lucida Grande",sans-serif;color:#222222'>Baldi,
P., &amp; Long, A Bayesian framework for the analysis of microarray expression
data: regularized t -test and statistical inferences of gene changes.
Bioinformatics, 17(6), 509–519 (2001).  </span></p>

<p class=RefText style='margin-left:39.7pt;line-height:normal'><span
style='font-size:11.0pt;font-family:"Lucida Grande",sans-serif;color:#222222'>&nbsp;</span></p>

<p class=RefText style='margin-left:39.7pt;line-height:normal'><span
style='font-size:11.0pt;font-family:"Lucida Grande",sans-serif;color:#222222'>Weile
et al., A framework for exhaustively mapping functional missense variants. Mol
Syst Biol&nbsp;13:&nbsp;957 (2017).</span></p>

<p class=RefText style='margin-left:39.7pt;line-height:normal'><span
style='font-size:11.0pt;font-family:"Lucida Grande",sans-serif;color:#222222'>&nbsp;</span></p>

<p class=RefText style='margin-left:39.7pt;line-height:normal'><span
style='font-size:11.0pt;font-family:"Lucida Grande",sans-serif;color:#222222'>Findlay,
G. M., Daza, R. M., Martin, B., Zhang, M. D., Leith, A. P., Gasperini, M., …
Shendure, J. (2018). Accurate functional classification of thousands of BRCA1
variants with saturation genome editing. BioRxiv.
https://doi.org/10.1101/294520</span></p>

<p style='margin-top:6.0pt;margin-right:0cm;margin-bottom:6.0pt;margin-left:
54.0pt;text-align:justify;text-justify:inter-ideograph;line-height:150%;
background:white'><span style='font-family:"Lucida Grande",sans-serif;
color:#222222'>&nbsp;</span></p>

</div>

</body>

</html>
